apps/guardian/src-tauri/src/main.rs

```diff
@@ -1029,7 +1046,363 @@ async fn generate_secret_share_flow<'a>(
+                Error::DecryptionShareError("Invalid decryption shares were detected".into())
+            })?;
+            Ok::<_, Error>(shares)
+        })
+        .try_collect()?;
+
+    /// The decryption data corresponding to a single [`Ciphertext`]
+    struct ProofData<'a> {
+        /// The ciphertext
+        ciphertext:        &'a Ciphertext,
+        /// The secret state for the commitment share for active guardian
+        secret_state:      &'a DecryptionProofStateShare,
+        /// A map of guardians and their decryption share for the ciphertext
+        decryption_shares: Vec<&'a DecryptionShareResult>,
+    }
+
+    // Generate the decryption proof for a single contest entry. An error is
+    // returned if either:
+    //
+    // - Shares received from peers cannot be combined
+    // - Decryption proof cannot be generated for some ciphertext
+    let generate_decryption_proof = |proof_data: &ProofData| -> Result<_, Error> {
+        let (commit_shares, decryption_shares): (Vec<_>, Vec<_>) = proof_data
+            .decryption_shares
+            .iter()
+            .map(|share| (share.proof_commit.clone(), &share.share))
+            .unzip();
+
+        let combined_decryption = CombinedDecryptionShare::combine(&parameters, decryption_shares)
+            .map_err(|e| {
+                eprintln!("{e}");
+                Error::DecryptionShareError("Failed to combine shares received from peers".into())
+            })?;
+        let proof = DecryptionProof::generate_response_share(
+            &parameters.fixed_parameters,
+            &context.hashes_ext,
+            &context.public_key,
+            proof_data.ciphertext,
+            &combined_decryption,
+            &commit_shares,
+            proof_data.secret_state,
+            &secret_key_share,
+        )
+        .map_err(|e| {
+            eprintln!("{e}");
+            Error::DecryptionShareError("Failed to generate the response share to register".into())
+        })?;
+
+        Ok(proof)
+    };
+
+    // A map of ciphertexts paired with associated secret state + guardian shares
+    // for each contest in the election. Errors are returned if:
+    //
+    // - Decryption shares do not match the format of the tally
+    let proof_data: BTreeMap<_, _> = encrypted_tally
+        .iter()
+        .map(|(contest_index, ciphertexts)| {
+            let secret_states = match secret_states.get(contest_index) {
+                Some(secret_states) if secret_states.len() == ciphertexts.len() => secret_states,
+                _ => return Err(anyhow!("Invalid secret states for tally")),
+            };
+            let decryption_shares: Vec<_> = decryption_shares
+                .iter()
+                .map(|shares| match shares.get(contest_index) {
+                    Some(shares) if shares.len() == ciphertexts.len() => Ok(shares),
+                    _ => Err(Error::DecryptionShareError(
+                        "Invalid decryption shares detected".into(),
```

Comment:
You could be a bit more precise here about what goes wrong. Might be useful for debugging/investigating.

Reply:
I'm deliberately trying to not be specific as the error message is meant to reach the end user and I reckon it just confuse if we're too specific. Also, the error returned from this branch might not provide the full picture anyway, but that is not the point anyway; we should just hint to the user that something is wrong, and they should contact the election coordinator. The intended consequence of this error should be to eliminate the guardian(s) at fault, at which point the election coordinator needs to be brought into play to figure out which guardians have shared something incorrect.